drivers/tsl4531x/tsl4531x.c

```diff
@@ -0,0 +1,180 @@
+                     TSL4531X_COMMAND(TSL4531X_ID_REG),
+                     &id, 0) < 0) {
+        DEBUG("[Error] Cannot read ID register.\n");
+        i2c_release(params->i2c_dev);
+        return -ENXIO;
+    }
+
+    DEBUG("[Info] tsl4531x sensor ID ? %d\n", id);
+
+    if (TSL4531X_GET_PARTNO(id) != params->part_number) {
+        DEBUG("[Error] not a TSL4531 sensor.\n");
+        i2c_release(params->i2c_dev);
+        return -ENOTSUP;
+    }
+
+    /* Configure device. In low power mode, we initially power the sensor down. */
+    if ((i2c_write_reg(params->i2c_dev,
+                       params->i2c_addr,
+                       TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                       TSL4531X_CONTROL(TSL4531X_POWER_MODE(params->low_power_mode)),
+                       0) < 0)
+        ||
+        (i2c_write_reg(params->i2c_dev,
+                       params->i2c_addr,
+                       TSL4531X_COMMAND(TSL4531X_CONFIGURATION_REG),
+                       TSL4531X_CONFIG(TSL4531X_PSAVESKIP_ON, params->integration_time),
+                       0) < 0)) {
+
+        DEBUG("[Error] Cannot configure device.\n");
+        i2c_release(params->i2c_dev);
+        return -ENXIO;
+    }
+
+    /* If device was configured correctly, initialise the device descriptor */
+    dev->i2c_dev = params->i2c_dev;
+    dev->i2c_addr = params->i2c_addr;
+    dev->integration_time = params->integration_time;
+    dev->low_power_mode = params->low_power_mode;
+    dev->sample_start_time = 0; /* Device assumed to start up at same time as
+                                   microcontroller - i.e. when it hits this line */
+
+    return 0;
+}
+
+int tsl4531x_set_low_power_mode(tsl4531x_t *dev, bool low_power_mode)
+{
+    assert(dev);
+
+    dev->low_power_mode = low_power_mode;
+
+    i2c_write_reg(dev->i2c_dev,
+                  dev->i2c_addr,
+                  TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                  TSL4531X_CONTROL(TSL4531X_POWER_MODE(low_power_mode)),
+                  0);
+
+    /* In high power mode only, we restart the sample ready timer, because only
+       in this mode it's used to indicate readiness after startup. */
+    if (!dev->low_power_mode) {
+        dev->sample_start_time = xtimer_now_usec();
+    }
+
+    return 0;
+}
+
+int tsl4531x_start_sample(tsl4531x_t *dev)
+{
+    assert(dev);
+
+    /* Don't change the mode to one-shot if the device is in high power mode. */
+    if (dev->low_power_mode) {
+        i2c_write_reg(dev->i2c_dev,
+                      dev->i2c_addr,
+                      TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                      TSL4531X_CONTROL(TSL4531X_MODE_SINGLE_ADC_CYCLE),
+                      0);
+
+        dev->sample_start_time = xtimer_now_usec();
+    }
+    else {
+        return -EPERM;
+    }
+
+    return 0;
+}
+
+bool tsl4531x_is_sample_ready(const tsl4531x_t *dev)
+{
+    assert(dev);
+
+    return ((xtimer_now_usec() - dev->sample_start_time) >
```

Comment:
As discussed offline, this will change anyway to the following functionality:
- Return time till sample ready
- Clamp sample ready time at zero if sample's ready